File: python/examples/by-use-case/dpsn-subscribe/llm_example.py
import os
import json
import time
from typing import Dict, Any, List
from dotenv import load_dotenv
from langchain_openai import ChatOpenAI
from langchain.agents import AgentExecutor, create_tool_calling_agent
from langchain_core.prompts import ChatPromptTemplate
from web3 import Web3

from goat_plugins.dpsn import dpsn
from goat_plugins.dpsn import DpsnPluginOptions
from goat_adapters.langchain import get_on_chain_tools
from goat_wallets.web3 import Web3EVMWalletClient


load_dotenv()
dpsn_url = os.getenv('DPSN_URL')
dpsn_pvt_key = os.getenv('EVM_WALLET_PVT_KEY')


BASE_RPC_URL = os.getenv("BASE_RPC_URL", "https://mainnet.base.org")
w3 = Web3(Web3.HTTPProvider(BASE_RPC_URL))
private_key = os.getenv("EVM_WALLET_PVT_KEY")
assert private_key is not None, "You must set WALLET_PRIVATE_KEY environment variable"
assert private_key.startswith("0x"), "Private key must start with 0x hex prefix"

# Global message store
# received_messages = []

def handle_dpsn_message(message: Dict[str, Any]):
    print("Received DPSN message:", message)
    return message

# Initialize LLM
llm = ChatOpenAI(model="gpt-4o-mini") # Use a specific model

def main():
      prompt = ChatPromptTemplate.from_messages(
        [
            ("system", """You are the Dpsn plugin tester. You can help users test DPSN functionality including:
1. Subscribe to a specific topic name using dpsn_subscribe.
   Example: "Subscribe to topic 0x123.../some/data"
2. Consume incoming messages on the stream and return the message            
3. Unsubscribe from a topic using dpsn_unsubscribe.
   Example: "Unsubscribe from topic 0x123.../some/data"

Remember to provide the full topic string when asked to subscribe or unsubscribe.
After subscribing, I'll wait for messages and include them in my response."""),
            ("placeholder", "{chat_history}"), # Keep chat history simple for this example
            ("human", "{input}"),
            ("placeholder", "{agent_scratchpad}"),
        ]
    )

      tools = get_on_chain_tools(
            wallet=Web3EVMWalletClient(w3),
            plugins=[
                dpsn(options=DpsnPluginOptions(dpsn_url,dpsn_pvt_key,handle_dpsn_message)) 
            ],
      )

      agent = create_tool_calling_agent(llm,tools,prompt)
      agent_executor = AgentExecutor(agent=agent, tools=tools, handle_parsing_errors=True, verbose=True)

      print("\n DPSN Plugin Test Interface")
      #You can select any topic from the DPSN Streams store
      #https://streams.dpsn.org
      print("Example Commands")
      print("1. Subscribe to   0xe14768a6d8798e4390ec4cb8a4c991202c2115a5cd7a6c0a7ababcaf93b4d2d4/BTCUSDT/ticker topic")
      print("2. Unsubscribe from 0xe14768a6d8798e4390ec4cb8a4c991202c2115a5cd7a6c0a7ababcaf93b4d2d4/BTCUSDT/ticker topic")

      while True:
  
           user_input = input("\n You: ").strip()

           if user_input.lower() == 'exit':
                print("goodbye!")
                break
           try:
                response = agent_executor.invoke({
                     "input": user_input
                })
                      
                print("\n Assistant:", response["output"])

           except Exception as e:
                print("\n Error:", str(e))


if __name__ == "__main__":
    main()

    

File: typescript/examples/by-use-case/dpsn-subscribe/llm-example.ts
// Create a .env file in this directory with:
// DPSN_URL=your_dpsn_url
// DPSN_PRIVATE_KEY=your_dpsn_private_key
// OPENAI_API_KEY=your_openai_key

import readline from "node:readline";
import { openai } from "@ai-sdk/openai";
import { Chain, WalletClientBase, getTools } from "@goat-sdk/core"; // Import necessary core components
import { ToolBase } from "@goat-sdk/core";
import { dpsnplugin } from "@goat-sdk/dpsn-plugin";
import { generateText } from "ai";
import * as dotenv from "dotenv";
import { ZodType, ZodTypeDef } from "zod";

// Define a minimal dummy wallet client for the example
// Replace this with your actual wallet setup (e.g., from @goat-sdk/wallet-evm)
class DummyWalletClient extends WalletClientBase {
    getAddress(): string {
        return "0xDummyAddress";
    }
    getChain(): Chain {
        // The DPSN plugin seems chain-agnostic, but core might require a chain.
        return { type: "evm", id: 1 }; // Example EVM chain
    }
    async signMessage(message: string): Promise<{ signature: string }> {
        console.warn("DummyWalletClient: signMessage called with:", message);
        return { signature: "0xDummySignature" };
    }
    async balanceOf(
        address: string,
    ): Promise<{ decimals: number; symbol: string; name: string; value: string; inBaseUnits: string }> {
        console.warn("DummyWalletClient: balanceOf called for:", address);
        return { decimals: 18, symbol: "ETH", name: "Ether", value: "0", inBaseUnits: "0" };
    }
}

// Load environment variables from .env file
dotenv.config();

async function setupAndRun() {
    // Use the dummy wallet
    const wallet = new DummyWalletClient();
    const dpsn_plugin = dpsnplugin({
        DPSN_URL: process.env.DPSN_URL ?? "",
        EVM_WALLET_PVT_KEY: process.env.EVM_WALLET_PVT_KEY ?? "",
    });
    //declare the dpsn data stream handler for listening to messages on subscribing to dpsn topics
    const DpsnDataStreamHandler = dpsn_plugin.DpsnDataStream;
    DpsnDataStreamHandler.on("message", (message: unknown) => {
        console.log("Received message from DPSN:", message);
        // Add your message processing logic here
    });
    // 1. Get tools array from the core wallet and the DPSN plugin
    const toolsArray = await getTools({
        // Renamed to toolsArray
        wallet: wallet,
        plugins: [
            dpsn_plugin, // Add the DPSN plugin
        ],
    });
    // Convert array to the object format expected by generateText (ToolSet)
    type ToolSet = { [key: string]: ToolBase<ZodType<unknown, ZodTypeDef, unknown>, unknown> };
    const toolSet: ToolSet = {};
    for (const tool of toolsArray) {
        // Assuming tool object has a 'name' property suitable for keying
        if (tool?.name) {
            toolSet[tool.name] = tool;
        }
    }

    console.log(
        "Tools retrieved:",
        Object.keys(toolSet), // Log keys from the new object
    );

    // 2. Create a readline interface to interact with the agent
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });

    console.log("\n--- Agent Setup Complete ---");
    console.log(
        "Enter prompts like 'subscribe to the 0xe14768a6d8798e4390ec4cb8a4c991202c2115a5cd7a6c0a7ababcaf93b4d2d4/BTCUSDT/ticker topic' or 'exit'",
    );
    //You can select any topic from the DPSN Streams store
    //https://streams.dpsn.org
    // 3. Start the interaction loop
    while (true) {
        const prompt = await new Promise<string>((resolve) => {
            rl.question('Enter your prompt (or "exit" to quit): ', resolve);
        });

        if (prompt.toLowerCase() === "exit") {
            rl.close();
            break;
        }

        console.log("\n-------------------");
        console.log("PROCESSING PROMPT...");
        console.log("\n-------------------");

        try {
            const { text, toolResults, toolCalls, steps } = await generateText({
                model: openai("gpt-4o-mini"),
                tools: toolSet, // Use the converted toolSet object
                maxSteps: 1,
                prompt: prompt,
            });

            // Check if toolResults exist and log them
            if (toolResults && toolResults.length > 0) {
                console.log("TOOL CALLED & RESULTS:");
                console.log(toolResults);

                // toolResults[0].result.on("message", (message: unknown) => {
                //     console.log("Received message from DPSN:", message);
                //     // Add your message processing logic here
                // });
            } else {
                console.log("NO TOOLS CALLED");
                console.log("\n-------------------");
            }
        } catch (error) {
            console.error("Error during generation:", error);
        }
    }
}

setupAndRun().catch(console.error);
